### DISCUSSION TITLE
`stderr`sink for `error`-level, `stdout` sink for other levels

### DISCUSSION THREAD

**Discussion by raphaelzulliger** [gabime/spdlog#1854] [Q&A] [answered]
It seems to me like a standard use case to use `stdout` and `stderr` sinks in an application and wanting that all error logs are going into the `stderr` sink, while all other go into the `stdout` sink.

I was not able to find any docs explaining that SpdLog supports "sinks per level" or the like. As a workaround, I can create two loggers in my code, an "error only"-logger and an "all other messages"-logger configure them with `stderr` and `stdout` sinks respectively and use the former for all error messages and the latter for the rest.

But since it looks to me like the "most standard use case", I feel like I'm doing something wrong here... Any hints?

Thanks!

**Comment by gabime** [comment, marked as answer]
spdlog supports setting level per sink

**Reply by raphaelzulliger** [reply]
Either there is a misunderstanding or I don't understand the proposed solution. Let me show my "problem" by an example. Assume we have:

```c++
    auto stdout_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
    stdout_sink->set_level(spdlog::level::trace);

    auto stderr_sink = std::make_shared<spdlog::sinks::stderr_color_sink_mt>();
    stderr_sink->set_level(spdlog::level::err);

    spdlog::logger logger("multi_sink", {stdout_sink, stderr_sink});
    // enable all logs
    logger.set_level(spdlog::level::trace);

    // NOK unfortunately, appears on stderr AND stdout:
    logger.error("ERROR");
    // OK: Only on stdout
    logger.warn("WARNING");
    // OK: Only on stdout
    logger.trace("TRACE");
```

The following happens:

```
xxx:~/spdlog_minimal$ ./build/bin/SpdLog_Minimal > /dev/null
[2021-03-04 10:18:40.020] [multi_sink] [error] ERROR
```

That's cool.

```
xxx:~/spdlog_minimal$ ./build/bin/SpdLog_Minimal 2> /dev/null
[2021-03-04 10:18:41.584] [multi_sink] [error] ERROR
[2021-03-04 10:18:41.584] [multi_sink] [warning] WARNING
[2021-03-04 10:18:41.584] [multi_sink] [trace] TRACE
```

This case is less cool. I'd say we rather want that "errors" are **only** on stderr, not on stdout. From a logical point of view, what I'd need is something like this:

```c++
// whatever the `set_level`, do *ignore* `spdlog::level::err`
stdout_sink->set_level_ignore(spdlog::level::err);
```

Do I have wrong expectations? Is it "normal" to output errors on `stdout` AND `stderr`?

**Reply by tt4g** [reply]
The spdlog level filter will pass all log messages above the configured level.
If you want to record only a specific log level, create a custom sync.

The following issues for your help.

* #1357
* #1408

**Reply by raphaelzulliger** [reply]
I see, that may help with my use case. Thx!
